hibernate-core/src/main/java/org/hibernate/mapping/SingleTableSubclass.java

```diff
@@ -40,4 +43,62 @@ public void validate(Metadata mapping) throws MappingException {
 		}
 		super.validate( mapping );
 	}
+
+	@Override
+	public void createConstraints(MetadataBuildingContext context) {
+		if ( !isAbstract() ) {
+			final Dialect dialect = context.getMetadataCollector().getDatabase().getDialect();
+			if ( dialect.supportsTableCheck() ) {
+				final List<Selectable> discriminator = getDiscriminator().getSelectables();
+				if ( discriminator.size() == 1
+						&& discriminator.get( 0 ) instanceof Column discriminatorColumn ) {
+					final StringBuilder check = new StringBuilder();
+					check.append( discriminatorColumn.getQuotedName( dialect ) );
```

Comment:
Why not support discriminator formulas as well?

```suggestion
					check.append( discriminator.get( 0 ).getText( dialect ) );
```

Reply:
I did try that and I forget the reason it didn’t quite work. I will try again.